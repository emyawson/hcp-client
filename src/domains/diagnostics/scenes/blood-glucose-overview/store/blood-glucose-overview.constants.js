import { translate } from 'src/i18n';

export const VARIABILITY_THRESHOLDS = {
  MEDIUM: 33,
  HIGH: 50,
};

export const TRAFFIC_LIGHT_COLORS = {
  GRAY: 'GRAY',
  GREEN: 'GREEN',
  RED: 'RED',
  YELLOW: 'YELLOW',
};

export const NO_MEASUREMENTS_INTERVAL_LABEL = '-';

export const TRAFFIC_LIGHT_LABELS = {
  HYPO: translate('bloodGlucoseOverview.hypo'),
  HYPER: translate('bloodGlucoseOverview.hyper'),
  IN_RANGE: translate('bloodGlucoseOverview.inRange'),
  LOW: translate('bloodGlucoseOverview.low'),
  MEDIUM: translate('bloodGlucoseOverview.medium'),
  HIGH: translate('bloodGlucoseOverview.high'),
  INSUFFICIENT_INFO: translate('bloodGlucoseOverview.na'),
  NO_INFO: NO_MEASUREMENTS_INTERVAL_LABEL,
};

export const MINIMUM_TESTS_FOR_INTERVAL = 28;
export const MINIMUM_AVERAGE_TESTS_FOR_DAY = 2;

export const NUMBER_BLOOD_GLUCOSE_OVERVIEW_COLUMNS = 6;

export const HYPO_RISK_THRESHOLDS = {
  MEDIUM: 1.1,
  HIGH: 2.5,
};
