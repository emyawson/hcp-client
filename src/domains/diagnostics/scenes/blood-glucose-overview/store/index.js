export {
  bloodGlucoseOverviewConnector,
  calculateAndFormatVariability,
  selectBloodGlucoseOverviewData,
  selectDayInHalfTimeIntervals,
  selectFormattedDateRanges,
  selectHypoglycaemia,
  selectHypoRisk,
  selectIsThereNextInterval,
  selectIsTherePrevInterval,
  selectMeanBloodGlucoseStatistics,
  selectMeanBloodGlucoseStatus,
  selectTestsPerDay,
  selectVerifiedGroupMeasurements,
} from './blood-glucose-overview.selector';
export {
  HYPO_RISK_MAX,
  HYPO_RISK_MIN,
  HYPO_RISK_THRESHOLDS,
  MIN_BG,
  MINIMUM_AVERAGE_TESTS_FOR_DAY,
  MINIMUM_TESTS_FOR_INTERVAL,
  NO_MEASUREMENTS_INTERVAL_LABEL,
  NUMBER_BLOOD_GLUCOSE_OVERVIEW_COLUMNS,
  TRAFFIC_LIGHT_COLORS,
  TRAFFIC_LIGHT_LABELS,
  VARIABILITY_THRESHOLDS,
} from './blood-glucose-overview.constants';
export {
  calculateAvgTestsPerDay,
  calculateBloodGlucoseIndex,
  calculateBloodGlucoseStartDate,
  calculateEndDateAfterPagination,
  calculateForVariability,
  calculateHighBloodGlucoseIndex,
  calculateHypoRisk,
  calculateHypoRiskColor,
  calculateHypoRiskForIntervals,
  calculateHypoRiskLabel,
  calculateIntervalEmptyDays,
  calculateLowBloodGlucoseIndex,
  calculateMeanBloodGlucose,
  calculateMeanColor,
  calculateMeanLabel,
  calculateVariabilityStatus,
  createCalculateForHypoglycaemia,
  determineHasReliableInfo,
  formatTestsPerDayAvg,
  groupByTimeInterval,
  hasReliableInfo,
  isFirstInterval,
  isHyperRed,
  isHyperYellow,
  isHypoRed,
  isHypoRiskHigh,
  isHypoRiskLow,
  isHypoRiskMedium,
  isHypoYellow,
  isInRange,
  isLastInterval,
  roundToNDecimalPlaces,
  toHypoglycaemiaCount,
  toHypoRiskStatus,
  toLabelInterval,
  toMeanBloodGlucoseValue,
  toTimeIntervals,
} from './blood-glucose-overview.utils';
