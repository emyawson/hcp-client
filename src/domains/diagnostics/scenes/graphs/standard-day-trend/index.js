export { StandardDayTrendContainer } from './standard-day-trend.container';
export { StandardDayTrend } from './standard-day-trend.component';
export {
  selectGraphData,
  selectIconHorizontalTicks,
  selectMealHorizontalTicks,
  selectTimeHorizontalTickLines,
  selectTimeHorizontalTicks,
  standardDayTrendConnector,
} from './standard-day-trend.selector';
