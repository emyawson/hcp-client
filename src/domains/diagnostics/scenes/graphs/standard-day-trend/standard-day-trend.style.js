import styled from 'styled-components';

import { spacing } from 'src/domains/diagnostics/styles';

import { MIN_EXPANDED_STD_GRAPH_HEIGHT_IN_REM } from '../graph.constants';

export const StandardDayTrendWrapperDiv = styled.div`
  position: relative;
  height: 100%;
  min-height: ${MIN_EXPANDED_STD_GRAPH_HEIGHT_IN_REM};
  margin: ${spacing.three} ${spacing.two} 0;
  display: flex;
  flex-direction: column;
`;

StandardDayTrendWrapperDiv.displayName = 'StandardDayTrendWrapperDiv';
