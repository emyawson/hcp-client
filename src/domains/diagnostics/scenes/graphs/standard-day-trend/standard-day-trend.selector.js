import { createSelector, createStructuredSelector } from 'reselect';

import { translate } from 'src/i18n'; // TODO: move to diagnostics
import {
  isDatetimeWithinInterval,
  getFormattedStandardDeviation,
} from 'src/domains/diagnostics/utils';
import {
  selectTargetRange,
  selectThreshold,
  selectGraphDetails,
  selectShowGridLines,
  selectVerticalTicks,
  selectVerticalAxesCeiling,
} from 'src/domains/diagnostics/scenes/graphs/graph.selector';
import {
  AppleIcon,
  AppleEatenIcon,
  NightIcon,
  OvernightIcon,
} from 'src/domains/diagnostics/assets/icons';
import {
  GRAPH_Y_MIN,
  GRAPH_Y_MAX,
} from 'src/domains/diagnostics/scenes/graphs/graph.constants';
import { EMPTY_VALUE_PLACEHOLDER } from 'src/domains/diagnostics/store/constants';
import {
  selectGlucoseMeasurementsInDateSliderRange,
  selectTimeIntervals,
  selectGraphLoading,
} from 'src/domains/diagnostics/store/selectors';

const getMeasurementDetails = measurements => {
  const glucoseValues = measurements.map(measurement => measurement.value);

  const max = Math.max(...glucoseValues);
  const min = Math.min(...glucoseValues);

  const sum = glucoseValues.reduce(
    (accumulator, value) => accumulator + value,
    0,
  );

  const mean = sum / glucoseValues.length;

  return {
    max,
    min,
    mean,
    count: glucoseValues.length,
    stdDev: getFormattedStandardDeviation(glucoseValues),
  };
};

const buildGlucoseDataSet = (measurements, intervals) =>
  intervals.map(interval =>
    getMeasurementDetails(
      measurements.filter(measurement =>
        isDatetimeWithinInterval(
          measurement.date,
          interval.startTime,
          interval.endTime,
        ),
      ),
    ),
  );

const normalizeGlucoseData = (data, floor, ceiling) =>
  data.map((datum, index) => {
    const { max, min, stdDev, mean } = datum;
    const bucketWidth = 1 / data.length;
    const x = index / data.length;
    const y = mean / ceiling;
    const mealIconTick = selectIconHorizontalTicks().find(
      iconTick => iconTick.value - bucketWidth / 2 === x,
    );
    return {
      max: max / ceiling,
      min: min / ceiling,
      deviation: stdDev === EMPTY_VALUE_PLACEHOLDER ? 0 : stdDev / ceiling,
      x,
      y,
      data: {
        ...datum,
        icon: mealIconTick ? mealIconTick.component : null, // Adding icon to data object represent icon for each specific bucket (used for tooltip)
      },
    };
  });

const normalizeGraphData = (measurements, intervals, graphYMax = GRAPH_Y_MAX) =>
  normalizeGlucoseData(
    buildGlucoseDataSet(measurements, intervals),
    GRAPH_Y_MIN,
    graphYMax,
  );

export const selectGraphData = createSelector(
  selectGlucoseMeasurementsInDateSliderRange,
  selectTimeIntervals,
  selectVerticalAxesCeiling,
  normalizeGraphData,
);

const normalizeTimeHorizontalTicks = (timeIntervals, index) =>
  timeIntervals.map((interval, index) => ({
    value: index / timeIntervals.length,
    label: interval.startTime.substr(0, interval.startTime.lastIndexOf(':')),
  }));

export const selectTimeHorizontalTicks = createSelector(
  selectTimeIntervals,
  normalizeTimeHorizontalTicks,
);

export const selectMealHorizontalTicks = () => [
  { value: 1 / 8, label: translate('general.mealBlocks.breakfast') },
  { value: 3 / 8, label: translate('general.mealBlocks.lunch') },
  { value: 5 / 8, label: translate('general.mealBlocks.dinner') },
  { value: 13 / 16, label: translate('general.mealBlocks.bedTime') },
  { value: 15 / 16, label: translate('general.mealBlocks.night') },
];

export const selectTimeHorizontalTickLines = () => [
  { value: 0, type: 'long' },
  { value: 1 / 8, type: 'short' },
  { value: 2 / 8, type: 'long' },
  { value: 3 / 8, type: 'short' },
  { value: 4 / 8, type: 'long' },
  { value: 5 / 8, type: 'short' },
  { value: 6 / 8, type: 'long' },
  { value: 7 / 8, type: 'short' },
  { value: 1, type: 'long' },
];

export const selectIconHorizontalTicks = () => [
  { value: 1 / 16, component: AppleIcon, iconWidthScale: 0.04 },
  { value: 3 / 16, component: AppleEatenIcon, iconWidthScale: 0.023 },
  { value: 5 / 16, component: AppleIcon, iconWidthScale: 0.04 },
  { value: 7 / 16, component: AppleEatenIcon, iconWidthScale: 0.023 },
  { value: 9 / 16, component: AppleIcon, iconWidthScale: 0.04 },
  { value: 11 / 16, component: AppleEatenIcon, iconWidthScale: 0.023 },
  { value: 13 / 16, component: NightIcon, iconWidthScale: 0.035 },
  { value: 15 / 16, component: OvernightIcon, iconWidthScale: 0.033 },
];

export const standardDayTrendConnector = createStructuredSelector({
  measurements: selectGlucoseMeasurementsInDateSliderRange,
  graphData: selectGraphData,
  targetRange: selectTargetRange,
  threshold: selectThreshold,
  verticalTicks: selectVerticalTicks,
  timeHorizontalTicks: selectTimeHorizontalTicks,
  timeHorizontalTickLines: selectTimeHorizontalTickLines,
  mealHorizontalTicks: selectMealHorizontalTicks,
  iconHorizontalTicks: selectIconHorizontalTicks,
  graphDetails: selectGraphDetails,
  showGridLines: selectShowGridLines,
  graphYMax: selectVerticalAxesCeiling,
  isLoading: selectGraphLoading,
});
