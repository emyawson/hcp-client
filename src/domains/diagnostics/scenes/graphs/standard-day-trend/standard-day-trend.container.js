import { compose } from 'recompose';
import { connect } from 'react-redux';

import { StandardDayTrend } from './standard-day-trend.component';
import { standardDayTrendConnector } from './standard-day-trend.selector';

export const StandardDayTrendContainer = compose(
  connect(
    standardDayTrendConnector,
    null,
  ),
)(StandardDayTrend);
