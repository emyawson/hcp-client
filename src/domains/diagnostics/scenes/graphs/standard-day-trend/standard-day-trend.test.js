import React from 'react';
import { shallow } from 'enzyme';
import { take } from 'ramda';

import {
  AppleIcon,
  AppleEatenIcon,
  NightIcon,
  OvernightIcon,
} from 'src/domains/diagnostics/assets/icons';
import {
  makeOverwrite,
  convertISOToJsGMT,
} from 'src/domains/diagnostics/utils';
import { EMPTY_VALUE_PLACEHOLDER } from 'src/domains/diagnostics/store/constants';

import {
  selectGraphData,
  selectTimeHorizontalTicks,
  selectMealHorizontalTicks,
  selectIconHorizontalTicks,
  standardDayTrendConnector,
} from './standard-day-trend.selector';
import { StandardDayTrend } from './standard-day-trend.component';

const mockedData = {
  stripDelivery: {
    timeIntervals: [
      {
        startTime: '05:00:00',
        endTime: '08:30:00',
      },
      {
        startTime: '08:30:00',
        endTime: '11:30:00',
      },
      {
        startTime: '11:30:00',
        endTime: '14:30:00',
      },
      {
        startTime: '14:30:00',
        endTime: '17:30:00',
      },
      {
        startTime: '17:30:00',
        endTime: '21:00:00',
      },
      {
        startTime: '21:00:00',
        endTime: '22:00:00',
      },
      {
        startTime: '22:00:00',
        endTime: '00:00:00',
      },
      {
        startTime: '00:00:00',
        endTime: '05:00:00',
      },
    ],
    threshold: {
      glucoseIdealIntervalMax: 125,
      glucoseIdealIntervalMin: 80,
      hypoglycemiaThreshold: 60,
    },
  },
  dashboard: {
    glucoseMeasurements: [
      {
        value: 212,
        date: convertISOToJsGMT('2017-07-01 07:37:00 GMT+0000'),
      },
      {
        value: 130,
        date: convertISOToJsGMT('2017-07-01 07:30:00 GMT+0000'),
      },
      {
        value: 79,
        date: convertISOToJsGMT('2017-07-02 09:35:00 GMT+0000'),
      },
      {
        value: 122,
        date: convertISOToJsGMT('2017-07-02 12:25:00 GMT+0000'),
      },
      {
        value: 142,
        date: convertISOToJsGMT('2017-07-02 14:55:00 GMT+0000'),
      },
      {
        value: 142,
        date: convertISOToJsGMT('2017-07-02 16:55:00 GMT+0000'),
      },
      {
        value: 222,
        date: convertISOToJsGMT('2017-07-02 20:55:00 GMT+0000'),
      },
      {
        value: 142,
        date: convertISOToJsGMT('2017-07-02 21:55:00 GMT+0000'),
      },
      {
        value: 179,
        date: convertISOToJsGMT('2017-07-02 23:35:00 GMT+0000'),
      },
      {
        value: 411,
        date: convertISOToJsGMT('2017-07-02 00:35:00 GMT+0000'),
      },
    ],
  },
};

const mockedState = {
  stripDelivery: {
    thresholds: {
      actualHyper: {
        preIdealInterval: 100,
        postIdealInterval: 99,
        noctIdealInterval: 99,
      },
      hyper: {
        preIdealInterval: 125,
        postIdealInterval: 99,
        noctIdealInterval: 99,
      },
      hypo: {
        preIdealInterval: 60,
        postIdealInterval: 99,
        noctIdealInterval: 99,
      },
      warning: {
        preIdealInterval: 80,
        postIdealInterval: 99,
        noctIdealInterval: 99,
      },
    },
    timeIntervals: [
      {
        id: 2161,
        description: 'BEFORE_BREAKFAST',
        startTime: '06:00:00',
        endTime: '08:30:00',
        label: null,
      },
      {
        id: 2162,
        description: 'AFTER_BREAKFAST',
        startTime: '08:30:00',
        endTime: '11:30:00',
        label: null,
      },
      {
        id: 2163,
        description: 'BEFORE_LUNCH',
        startTime: '11:30:00',
        endTime: '12:30:00',
        label: null,
      },
      {
        id: 2164,
        description: 'AFTER_LUNCH',
        startTime: '12:30:00',
        endTime: '14:30:00',
        label: null,
      },
      {
        id: 2165,
        description: 'BEFORE_DINNER',
        startTime: '14:30:00',
        endTime: '18:00:00',
        label: null,
      },
      {
        id: 2166,
        description: 'AFTER_DINNER',
        startTime: '18:00:00',
        endTime: '21:00:00',
        label: null,
      },
      {
        id: 2167,
        description: 'BEDTIME',
        startTime: '21:00:00',
        endTime: '01:00:00',
        label: null,
      },
      {
        id: 2168,
        description: 'NIGHT',
        startTime: '01:00:00',
        endTime: '06:00:00',
        label: null,
      },
    ],
  },
  ui: {
    patientDashboard: {
      bloodGlucoseUnit: 'mg/dL',
      glucoseMeasurements: [
        {
          value: 212,
          date: convertISOToJsGMT('Wed Nov 08 2017 18:39:25 GMT-0500 (EST)'),
        },
        {
          value: 130,
          date: convertISOToJsGMT('Wed Nov 09 2017 13:39:25 GMT-0500 (EST)'),
        },
        {
          value: 140,
          date: convertISOToJsGMT('Wed Nov 09 2017 20:39:25 GMT-0500 (EST)'),
        },
      ],
      graphStartTime: '0:00',
      showBloodGlucoseLines: true,
      showBloodGlucosePoints: true,
      showBloodGlucoseBeforeMealPoints: true,
      showBloodGlucoseAfterMealPoints: true,
      showCarbohydrates: true,
      showMeanBloodGlucose: true,
    },
    patientDateRange: {
      startDate: convertISOToJsGMT('Mon Nov 06 2017 15:55:53 GMT+0000'),
      endDate: convertISOToJsGMT('Thu Nov 09 2017 23:59:33 GMT+0000'),
    },
  },
};

const overwriteMockedState = makeOverwrite(mockedState);

describe('Standard Day Trend Component', () => {
  it('renders correctly', () => {
    const mockProps = {
      graphData: [
        {
          max: 0.605,
          min: 0.155,
          deviation: 0.23340326333051414,
          x: 0,
          y: 0.41583333333333333,
          data: {
            max: 242,
            min: 62,
            mean: 166.33333333333334,
            stdDev: 93.36130533220566,
          },
        },
      ],
      targetRange: {
        minTargetWindow: 0.245,
        maxTargetWindow: 0.32,
      },
      threshold: 0.22,
    };

    const wrapper = shallow(<StandardDayTrend {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Standard Day Trend Selector', () => {
  it('should normalize graph data', () => {
    const { timeIntervals } = mockedData.stripDelivery;
    const { glucoseMeasurements } = mockedData.dashboard;

    expect(
      selectGraphData.resultFunc(glucoseMeasurements, timeIntervals),
    ).toEqual([
      {
        data: {
          count: 2,
          icon: AppleIcon,
          max: 212,
          mean: 171,
          min: 130,
          stdDev: 57.982756057296896,
        },
        deviation: 0.14495689014324223,
        max: 0.53,
        min: 0.325,
        x: 0,
        y: 0.4275,
      },
      {
        data: {
          count: 1,
          icon: AppleEatenIcon,
          max: 79,
          mean: 79,
          min: 79,
          stdDev: EMPTY_VALUE_PLACEHOLDER,
        },
        deviation: 0,
        max: 0.1975,
        min: 0.1975,
        x: 0.125,
        y: 0.1975,
      },
      {
        data: {
          count: 1,
          icon: AppleIcon,
          max: 122,
          mean: 122,
          min: 122,
          stdDev: EMPTY_VALUE_PLACEHOLDER,
        },
        deviation: 0,
        max: 0.305,
        min: 0.305,
        x: 0.25,
        y: 0.305,
      },
      {
        data: {
          count: 2,
          icon: AppleEatenIcon,
          max: 142,
          mean: 142,
          min: 142,
          stdDev: 0,
        },
        deviation: 0,
        max: 0.355,
        min: 0.355,
        x: 0.375,
        y: 0.355,
      },
      {
        data: {
          count: 1,
          icon: AppleIcon,
          max: 222,
          mean: 222,
          min: 222,
          stdDev: EMPTY_VALUE_PLACEHOLDER,
        },
        deviation: 0,
        max: 0.555,
        min: 0.555,
        x: 0.5,
        y: 0.555,
      },
      {
        data: {
          count: 1,
          icon: AppleEatenIcon,
          max: 142,
          mean: 142,
          min: 142,
          stdDev: EMPTY_VALUE_PLACEHOLDER,
        },
        deviation: 0,
        max: 0.355,
        min: 0.355,
        x: 0.625,
        y: 0.355,
      },
      {
        data: {
          count: 1,
          icon: NightIcon,
          max: 179,
          mean: 179,
          min: 179,
          stdDev: EMPTY_VALUE_PLACEHOLDER,
        },
        deviation: 0,
        max: 0.4475,
        min: 0.4475,
        x: 0.75,
        y: 0.4475,
      },
      {
        data: {
          count: 1,
          icon: OvernightIcon,
          max: 411,
          mean: 411,
          min: 411,
          stdDev: EMPTY_VALUE_PLACEHOLDER,
        },
        deviation: 0,
        max: 1.0275,
        min: 1.0275,
        x: 0.875,
        y: 1.0275,
      },
    ]);
  });

  it('should select time horizontal graph tick values', () => {
    const { timeIntervals } = mockedData.stripDelivery;

    expect(selectTimeHorizontalTicks.resultFunc(timeIntervals)).toEqual([
      { label: '05:00', value: 0 },
      { label: '08:30', value: 0.125 },
      { label: '11:30', value: 0.25 },
      { label: '14:30', value: 0.375 },
      { label: '17:30', value: 0.5 },
      { label: '21:00', value: 0.625 },
      { label: '22:00', value: 0.75 },
      { label: '00:00', value: 0.875 },
    ]);
  });

  it('should select the correct icon horizontal graph ticks', () => {
    expect(selectIconHorizontalTicks()).toEqual([
      { value: 1 / 16, component: AppleIcon, iconWidthScale: 0.04 },
      { value: 3 / 16, component: AppleEatenIcon, iconWidthScale: 0.023 },
      { value: 5 / 16, component: AppleIcon, iconWidthScale: 0.04 },
      { value: 7 / 16, component: AppleEatenIcon, iconWidthScale: 0.023 },
      { value: 9 / 16, component: AppleIcon, iconWidthScale: 0.04 },
      { value: 11 / 16, component: AppleEatenIcon, iconWidthScale: 0.023 },
      { value: 13 / 16, component: NightIcon, iconWidthScale: 0.035 },
      { value: 15 / 16, component: OvernightIcon, iconWidthScale: 0.033 },
    ]);
  });

  it('should select meal horizontal graph tick values', () => {
    expect(selectMealHorizontalTicks()).toEqual([
      { value: 1 / 8, label: 'Breakfast' },
      { value: 3 / 8, label: 'Lunch' },
      { value: 5 / 8, label: 'Dinner' },
      { value: 13 / 16, label: 'Bed Time' },
      { value: 15 / 16, label: 'Night' },
    ]);
  });

  describe('dynamic vertical ticks', () => {
    describe('when the highest measurements value is less than 400', () => {
      const glucoseMeasurements = [
        {
          date: convertISOToJsGMT('Wed Nov 08 2017 18:39:25 GMT-0500 (EST)'),
          beforeMeal: true,
          afterMeal: false,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 230,
        },
        {
          date: convertISOToJsGMT('Thu Nov 08 2017 15:39:25 GMT-0500 (EST)'),
          beforeMeal: false,
          afterMeal: true,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 240,
        },
        {
          date: convertISOToJsGMT('Thu Nov 09 2017 16:39:25 GMT-0500 (EST)'),
          beforeMeal: false,
          afterMeal: true,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 280,
        },
      ];

      it('vertical ticks start from highest value rounded up to a dividend of 50', () => {
        const expected = [
          { gridLine: true, label: '0', value: 0 },
          { gridLine: true, label: '50', value: 50 / 300 },
          { gridLine: true, label: '100', value: 100 / 300 },
          { gridLine: true, label: '150', value: 150 / 300 },
          { gridLine: true, label: '200', value: 200 / 300 },
          { gridLine: true, label: '250', value: 250 / 300 },
          { gridLine: true, label: '300', value: 1 },
        ];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = take(7)(
          standardDayTrendConnector(mockState).verticalTicks,
        );
        expect(actual).toEqual(expected);
      });

      it('graph data should reflect the dynamic ceiling', () => {
        const afterBed = 240 / 300;
        const dinner = (280 + 230) / 2 / 300;
        const expected = [NaN, NaN, NaN, NaN, NaN, afterBed, dinner, NaN];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = standardDayTrendConnector(mockState).graphData.map(
          x => x.y,
        );
        expect(actual).toEqual(expected);
      });

      it('target range should reflect the dynamic ceiling', () => {
        const warningY = 80 / 300;
        const hyperY = 125 / 300;
        const expected = { min: warningY, max: hyperY };

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = standardDayTrendConnector(mockState).targetRange;
        expect(actual).toEqual(expect.objectContaining(expected));
      });

      it('threshold should reflect the dynamic ceiling', () => {
        const expected = 60 / 300;

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = standardDayTrendConnector(mockState).threshold.value;
        expect(actual).toEqual(expected);
      });
    });

    describe('when the highest measurements value is less than the upper limit', () => {
      const glucoseMeasurements = [
        {
          date: convertISOToJsGMT('Wed Nov 08 2017 18:39:25 GMT-0500 (EST)'),
          beforeMeal: true,
          afterMeal: false,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 20,
        },
        {
          date: convertISOToJsGMT('Thu Nov 08 2017 15:39:25 GMT-0500 (EST)'),
          beforeMeal: true,
          afterMeal: false,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 24,
        },
        {
          date: convertISOToJsGMT('Thu Nov 09 2017 16:39:25 GMT-0500 (EST)'),
          beforeMeal: false,
          afterMeal: true,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 30,
        },
      ];

      it('vertical ticks start from the dividend of 50 above upper limit', () => {
        const expected = [
          { gridLine: true, label: '0', value: 0 },
          { gridLine: true, label: '50', value: 1 / 3 },
          { gridLine: true, label: '100', value: 2 / 3 },
          { gridLine: true, label: '150', value: 3 / 3 },
        ];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = take(4)(
          standardDayTrendConnector(mockState).verticalTicks,
        );
        expect(actual).toEqual(expected);
      });

      it('graph data should reflect the dynamic ceiling', () => {
        const afterBed = 24 / 150;
        const dinner = (30 + 20) / 2 / 150;
        const expected = [NaN, NaN, NaN, NaN, NaN, afterBed, dinner, NaN];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = standardDayTrendConnector(mockState).graphData.map(
          x => x.y,
        );
        expect(actual).toEqual(expected);
      });

      it('target range should reflect the dynamic ceiling', () => {
        const warningY = 80 / 150;
        const hyperY = 125 / 150;
        const expected = { min: warningY, max: hyperY };

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = standardDayTrendConnector(mockState).targetRange;
        expect(actual).toEqual(expect.objectContaining(expected));
      });

      it('threshold should reflect the dynamic ceiling', () => {
        const expected = 60 / 150;

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = standardDayTrendConnector(mockState).threshold.value;
        expect(actual).toEqual(expected);
      });
    });

    describe('when the highest measurements value is more than 400', () => {
      const glucoseMeasurements = [
        {
          date: convertISOToJsGMT('Wed Nov 08 2017 18:39:25 GMT-0500 (EST)'),
          beforeMeal: true,
          afterMeal: false,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 230,
        },
        {
          date: convertISOToJsGMT('Thu Nov 08 2017 15:39:25 GMT-0500 (EST)'),
          beforeMeal: false,
          afterMeal: true,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 240,
        },
        {
          date: convertISOToJsGMT('Thu Nov 09 2017 16:39:25 GMT-0500 (EST)'),
          beforeMeal: false,
          afterMeal: true,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 420,
        },
      ];

      it('vertical ticks start from 400', () => {
        const expected = [
          { gridLine: true, label: '0', value: 0 },
          { gridLine: true, label: '50', value: 50 / 400 },
          { gridLine: true, label: '100', value: 100 / 400 },
          { gridLine: true, label: '150', value: 150 / 400 },
          { gridLine: true, label: '200', value: 200 / 400 },
          { gridLine: true, label: '250', value: 250 / 400 },
          { gridLine: true, label: '300', value: 300 / 400 },
          { gridLine: true, label: '350', value: 350 / 400 },
          { gridLine: true, label: '400', value: 1 },
        ];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = take(9)(
          standardDayTrendConnector(mockState).verticalTicks,
        );
        expect(actual).toEqual(expected);
      });

      it('graph data should reflect the dynamic ceiling', () => {
        const afterBed = 240 / 400;
        const dinner = (420 + 230) / 2 / 400;
        const expected = [NaN, NaN, NaN, NaN, NaN, afterBed, dinner, NaN];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = standardDayTrendConnector(mockState).graphData.map(
          x => x.y,
        );
        expect(actual).toEqual(expected);
      });

      it('target range should reflect the dynamic ceiling', () => {
        const warningY = 80 / 400;
        const hyperY = 125 / 400;
        const expected = { min: warningY, max: hyperY };

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = standardDayTrendConnector(mockState).targetRange;
        expect(actual).toEqual(expect.objectContaining(expected));
      });

      it('threshold should reflect the dynamic ceiling', () => {
        const expected = 60 / 400;

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = standardDayTrendConnector(mockState).threshold.value;
        expect(actual).toEqual(expected);
      });
    });
  });
});
