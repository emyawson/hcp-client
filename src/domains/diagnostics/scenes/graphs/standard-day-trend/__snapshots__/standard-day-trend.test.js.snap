// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Standard Day Trend Component renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withState(withHandlers(Component))
    graphData={
      Array [
        Object {
          "data": Object {
            "max": 242,
            "mean": 166.33333333333334,
            "min": 62,
            "stdDev": 93.36130533220566,
          },
          "deviation": 0.23340326333051414,
          "max": 0.605,
          "min": 0.155,
          "x": 0,
          "y": 0.41583333333333333,
        },
      ]
    }
    targetRange={
      Object {
        "maxTargetWindow": 0.32,
        "minTargetWindow": 0.245,
      }
    }
    threshold={0.22}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "class",
    "props": Object {
      "graphData": Array [
        Object {
          "data": Object {
            "max": 242,
            "mean": 166.33333333333334,
            "min": 62,
            "stdDev": 93.36130533220566,
          },
          "deviation": 0.23340326333051414,
          "max": 0.605,
          "min": 0.155,
          "x": 0,
          "y": 0.41583333333333333,
        },
      ],
      "targetRange": Object {
        "maxTargetWindow": 0.32,
        "minTargetWindow": 0.245,
      },
      "threshold": 0.22,
      "toggleToolTip": [Function],
      "toolTip": Object {
        "data": Object {},
        "x": null,
        "y": null,
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "graphData": Array [
          Object {
            "data": Object {
              "max": 242,
              "mean": 166.33333333333334,
              "min": 62,
              "stdDev": 93.36130533220566,
            },
            "deviation": 0.23340326333051414,
            "max": 0.605,
            "min": 0.155,
            "x": 0,
            "y": 0.41583333333333333,
          },
        ],
        "targetRange": Object {
          "maxTargetWindow": 0.32,
          "minTargetWindow": 0.245,
        },
        "threshold": 0.22,
        "toggleToolTip": [Function],
        "toolTip": Object {
          "data": Object {},
          "x": null,
          "y": null,
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
